synapse/storage/devices.py

```diff
@@ -557,6 +557,9 @@ def _prune_txn(txn):
             txn.execute(select_sql)
             rows = txn.fetchall()
 
+            if not rows:
+                return
```

Comment:
I wouldn't expect `min(ts) < ?` to filter out many rows though, as most users won't update particularly frequently.